cmd/prometheus/main.go

```diff
@@ -450,17 +465,14 @@ func main() {
 		)
 	}
 	{
-		// TODO(krasi) refactor targetManager.Run() to be blocking to avoid using an extra blocking channel.
-		cancel := make(chan struct{})
 		g.Add(
 			func() error {
-				targetManager.Run()
-				<-cancel
-				return nil
+				return scrapeManager.Run(discoveryManager.SyncCh())
 			},
 			func(err error) {
-				targetManager.Stop()
-				close(cancel)
+				level.Info(logger).Log("msg", "Stopping scrape manager...")
+				cancelCtx()
+				level.Info(logger).Log("msg", "Scrape manager stopped")
```

Comment:
Same, this should be a defer in the run func above.

The first log line should probably also go there. At least from my understanding this function here should only send initial signal(s) for shutdown to the main runner func.

Reply:
yeas, I overlooked these parts a bit.
if you don't mind for such a small block I prefer that I do it without a `defer` for a more natural flow look.

```
func() error {
	err := scrapeManager.Run(discoveryManager.SyncCh())
	level.Info(logger).Log("msg", "Scrape manager stopped")
	return err
	},
func(err error) {
	level.Info(logger).Log("msg", "Stopping scrape manager...")
	cancelCtx()
	},
```

by the way I don't see how can we put the `Stopping scrape manager...` in the first block?
we need this log to show as soon as we start the cancelation and that can't happen with the blocking `scrapeManager.Run` function.